Keypad/Keypad.js

```diff
@@ -0,0 +1,327 @@
+							shrink
+							subtext={keyText.subtext}
+							text={keyText.text === 'arrowleftturn' || keyText.text === 'phone' ? null : keyText.text}
+						>
+							{keyText.text === 'arrowleftturn' || keyText.text === 'phone' ? keyText.text : null}
+						</Cell>
+					);
+				})}
+			</Layout>
+		);
+	}
+});
+
+/**
+ * A Keypad component with an Input to display the outcome.
+ *
+ * @class Keypad
+ * @memberof agate/Keypad
+ * @extends agate/Keypad.KeypadBase
+ * @ui
+ * @public
+ */
+class Keypad extends React.Component {
+	static propTypes = /** @lends agate/Keypad.Keypad.prototype */ {
+		/**
+		 * Applies a disabled style and the control becomes non-interactive.
+		 *
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		disabled: PropTypes.bool,
+
+		/**
+		 * Called when a button is clicked. Includes the 'key' key in its event payload, updates the state and the input value accordingly.
+		 *
+		 * @type {Function}
+		 * @param {Object} event
+		 * @public
+		 */
+		handleInputValue: PropTypes.func,
+
+		/**
+		 * Called when the input value is changed.
+		 *
+		 * @type {Function}
+		 * @param {Object} event
+		 * @public
+		 */
+		onChange: PropTypes.func,
+
+		/**
+		 * The value of the input.
+		 *
+		 * @type {String}
+		 * @public
+		 */
+		value: PropTypes.string
+	}
+
+	constructor (props) {
+		super(props);
+
+		this.state = {
+			charIndex: 0,
+			keypadInput: ''
+		};
+	}
+
+	handleInputValue = (keyValue) => {
+		const {charIndex, keypadInput} = this.state;
+		let newKeypadInput = keypadInput;
+		let newCharIndex;
+
+		switch (keyValue) {
+			case 'arrowleftturn':
+			case 'Backspace':
+				newCharIndex = charIndex;
+				newKeypadInput = newKeypadInput.substring(0, charIndex - 1) + newKeypadInput.substring(charIndex, newKeypadInput.length);
+				newCharIndex = newCharIndex - 1;
+
+				this.setState({
+					charIndex: newCharIndex
+				});
+				break;
+
+			case 'ArrowLeft':
+				if (charIndex >= 0) {
+					newCharIndex = charIndex;
+					newCharIndex = newCharIndex - 1;
+
+					this.setState({
+						charIndex: newCharIndex
+					});
+				}
+				break;
+
+			case 'ArrowRight':
+				newCharIndex = charIndex;
+				newCharIndex = newCharIndex + 1;
+
+				this.setState({
+					charIndex: newCharIndex
+				});
+				break;
+
+			case 'Delete':
+				newCharIndex = charIndex;
+				newKeypadInput = newKeypadInput.substring(0, charIndex) + newKeypadInput.substring(charIndex + 1, newKeypadInput.length);
+				newCharIndex = newCharIndex - 1;
+
+				this.setState({
+					charIndex: newCharIndex
+				});
+				break;
+
+			case 'ArrowUp':
+			case 'ArrowDown':
+				// do nothing;
+				break;
+
+			case 'phone':
+				// method to call dialed number (keypadInput);
+				break;
+
+			default:
+				newCharIndex = charIndex;
+				newKeypadInput = newKeypadInput.substring(0, charIndex) + keyValue + newKeypadInput.substring(charIndex);
+				newCharIndex = newCharIndex + 1;
+
+				this.setState({
+					charIndex: newCharIndex
+				});
```

Comment:
It looks like you're calling multiple setStates in a single function, but also firing an onChange callback before the final state is set. What if we set `charIndex` in the last setState, and in this switch statement didn't issue any setStates?